File: javascript/src/GroupBuilder.ts
import { RegExpExecArray } from 'regexp-match-indices'

import Group from './Group.js'

export default class GroupBuilder {
  public source: string
  public capturing = true
  private readonly groupBuilders: GroupBuilder[] = []

  public add(groupBuilder: GroupBuilder) {
    this.groupBuilders.push(groupBuilder)
  }

  public build(match: RegExpExecArray, nextGroupIndex: () => number): Group | null {
    const groupIndex = nextGroupIndex()
    const children = this.groupBuilders.map((gb) => gb.build(match, nextGroupIndex))
    const value = match[groupIndex]
    if (value === undefined) return null
    const index = match.indices[groupIndex]
    const start = index[0]
    const end = index[1]
    return new Group(value, start, end, children)
  }

  public setNonCapturing() {
    this.capturing = false
  }

  get children() {
    return this.groupBuilders
  }

  public moveChildrenTo(groupBuilder: GroupBuilder) {
    this.groupBuilders.forEach((child) => groupBuilder.add(child))
  }
}

File: dart/lib/src/group_builder.dart
import 'package:cucumber_expressions/src/group.dart';

class GroupBuilder {
  String source = '';

  bool capturing = true;
  final List<GroupBuilder> _groupBuilders = <GroupBuilder>[];

  void add(GroupBuilder groupBuilder) {
    _groupBuilders.add(groupBuilder);
  }

  Group build(RegExpMatch match, int Function() nextGroupIndex) {
    final groupIndex = nextGroupIndex();
    final children =
        _groupBuilders.map((gb) => gb.build(match, nextGroupIndex)).toList();
    final value = match.group(groupIndex);
    int? start;
    int? end;
    if (value != null) {
      start = match.start;
      end = match.end;
    }
    return buildGroup(
      value,
      groupIndex == 0 ? start : null,
      groupIndex == 0 ? end : null,
      children.isEmpty ? null : children,
    );
  }

  void setNonCapturing() {
    capturing = false;
  }

  List<GroupBuilder> get children => _groupBuilders;

  void moveChildrenTo(GroupBuilder groupBuilder) {
    _groupBuilders.forEach(groupBuilder.add);
  }
}
